[CodeCompletion] Rework getOperatorCompletions()

rintaro opened the pull request:
In postfix completion, for operator completion, we do:

  1. Type check the operand without applying it, but set the resolved type to the root of the expression.
  2. For each possible operators:
      1. Build temporary binary/postfix expression
      2. Perform type checking to see whether the operator is applicable 

This could be very slow especially if the operand is complex.

* Introduce `ReusePrecheckedType` option to constraint system. With this option, CSGen respects pre-stored types in expressions and doesn't take its sub-expressions into account.
  * Improve type checking performance because type variables aren't generated for sub-expressions of LHS (45511835)
  * Guarantee that the operand is not modified by the type checker because expression walkers in `CSGen` doesn't walk into the operand.

* Introduce `TypeChecker::findLHS()` to find LHS for a infix operator from pre-folded expression. We used to `foldSequence()` temporary `SequenceExpr` and find 'CodeCompletionExpr' for each attempt.
  * No need to flatten folded expression after initial type-checking.
  * Save memory of temporary `BinaryExpr` which used to be allocated by `foldSequence()`.
  * Improve accuracy of the completion. `foldSequence()` recovers invalid combination of operators by `left` associative manner (with diagnostics). This used to cause false-positive results. For instance, `a == b <HERE>` used to suggest `==` operator. `findLHS()` returns `nullptr` for such invalid combination.

rdar://problem/45511835
https://bugs.swift.org/browse/SR-9061

rintaro committed:
[CodeCompletion] Rework getOperatorCompletions()

* Introduce `FixedTypeExpr` as the LHS for type checking attempt.
  * Guarantees that the actual LHS is not modified by the type checker.
  * Improves type checking performance because it doesn't have
    any sub-expressions. (45511835)

* Introduce `TypeChecker::findLHS()` to find LHS for a infix operator
  from pre-folded expression. We used to `foldSequence()` temporary
  `SequenceExpr` and find 'CodeCompletionExpr' for each attempt.
  * No need to flatten folded expression after initial type-checking.
  * Save memory of temporary `BinaryExpr` which used to be allocated by
    `foldSequence()`.
  * Improve accuracy of the completion. `foldSequence()` recovers invalid
    combination of operators by `left` associative manner (with
    diagnostics). This used to cause false-positive results. For instance,
    `1 == 1 <HERE>` used to suggest `==` operator. `findLHS()` returns
    `nullptr` for such invalid combination.

rdar://problem/45511835
https://bugs.swift.org/browse/SR-9061

rintaro commented:
@swift-ci Please smoke test

xedin commented on include/swift/Sema/IDETypeChecking.h:
```diff
@@ -98,19 +98,15 @@ namespace swift {
+  /// or \c DeclRefKind::BinaryOperator for infix operator.
+  /// On success, returns resolved function type of the operator. The LHS should
+  /// already be type-checked. This function guarantees LHS not to be modified.
+  Type getTypeOfCompletionOperator(DeclContext *DC, Expr *LHS,
```
Wouldn't this strictly be a `FunctionType`? maybe it makes sense to return it directly.

xedin reviewed (commented):
LGTM! But I wonder if it's possible to avoid dummy expression, instead generate constraints only for   Apply + RHS and just cache types of the left-hand side in the constraint system (because that sub-tree is already type-checked), since you are already using `solve` that doesn't call `ExprRewriter` this might be simpler solution for the problem at hand.

rintaro committed:
[CodeCompletion] Add constraint system option to reuse pre-checked type

rintaro committed:
[CodeCompletion] Return FunctionType directly from getTypeOfCompletionOperator

There's no reason to pack it into 'Type'.

rintaro commented:
@xedin Thanks! I updated PR. Added constraint system flag `ReusePrecheckedType` to reuse LHS type and not to take its subexpressions into account. What do you think?

rintaro committed:
[CodeCompletion] Add fixed test case for code completion slowness

rintaro commented:
@swift-ci Please smoke test

xedin reviewed (approved):
Yeah, I think new approach makes sense! One more thing - I think it would be fair to assert that expressions don't have types with type variables in them, because solver isn't supposed to set types like that to AST.

rintaro commented:
@xedin Thanks!

> it would be fair to assert that expressions don't have types with type variables in them

Sorry to ask a dumb question but what is "them" do you mean? input `LHS`?

xedin commented:
Ah no problem! I mean expressions should always either have a valid type or unresolved type, because CSGen never sets types on expressions directly anymore, but rather uses a separate cache associated with each constraint system to hold intermisiate types with type variables.

rintaro committed:
[CSGen] Assert expression doesn't have type variables in ReusePrecheckedType mode

rintaro commented:
Thanks for the explanation. I added assertion in CSGen.
@swift-ci Please smoke test
